arch/rv32i/src/clic.rs

```diff
@@ -0,0 +1,299 @@
+            if pending.is_set(intpend::IntPend)
+                && self.registers.clicintie.localint[i].is_set(inten::IntEn)
+            {
+                self.registers.clicintie.localint[i].write(inten::IntEn::CLEAR);
+            }
+        }
+    }
+
+    /// Disable all interrupts.
+    pub fn disable_all(&self) {
+        self.registers.clicintie.msip.write(inten::IntEn::CLEAR);
+        self.registers.clicintie.mtip.write(inten::IntEn::CLEAR);
+        self.registers.clicintie.meip.write(inten::IntEn::CLEAR);
+        self.registers.clicintie.csip.write(inten::IntEn::CLEAR);
+
+        for enable in self.registers.clicintie.localint.iter() {
+            enable.write(inten::IntEn::CLEAR);
+        }
+    }
+
+    /// Get the index (0-256) of the lowest number pending interrupt, or `None` if
+    /// none is pending.
+    pub fn next_pending(&self) -> Option<u32> {
+        if self.in_use_interrupts & (1 << 3) > 0
+            && self.registers.clicintip.msip.is_set(intpend::IntPend)
+        {
+            return Some(3);
+        } else if self.in_use_interrupts & (1 << 7) > 0
+            && self.registers.clicintip.mtip.is_set(intpend::IntPend)
+        {
+            return Some(7);
+        } else if self.in_use_interrupts & (1 << 11) > 0
+            && self.registers.clicintip.meip.is_set(intpend::IntPend)
+        {
+            return Some(11);
+        } else if self.in_use_interrupts & (1 << 12) > 0
+            && self.registers.clicintip.csip.is_set(intpend::IntPend)
+        {
+            return Some(12);
+        }
+
+        for (i, pending) in self.registers.clicintip.localintpend.iter().enumerate() {
+            if self.in_use_interrupts & (1 << (i + 16)) > 0 && pending.is_set(intpend::IntPend) {
+                return Some((i + 16) as u32);
+            }
+        }
+        return None;
+    }
+
+    /// Signal that an interrupt is finished being handled. In Tock, this should
+    /// be called from the normal main loop (not the interrupt handler). This
+    /// marks the interrupt as no longer pending and re-enables it.
+    pub fn complete(&self, index: u32) {
+        match index {
+            3 => {
+                self.registers.clicintip.msip.write(intpend::IntPend::CLEAR);
+                self.registers.clicintie.msip.write(inten::IntEn::SET);
+            }
+            7 => {
+                self.registers.clicintip.mtip.write(intpend::IntPend::CLEAR);
+                self.registers.clicintie.mtip.write(inten::IntEn::SET);
+            }
+            11 => {
+                self.registers.clicintip.meip.write(intpend::IntPend::CLEAR);
+                self.registers.clicintie.meip.write(inten::IntEn::SET);
+            }
+            12 => {
+                self.registers.clicintip.csip.write(intpend::IntPend::CLEAR);
+                self.registers.clicintie.csip.write(inten::IntEn::SET);
+            }
+            16...144 => {
```

Comment:
range patterns will be deprecated soon

```suggestion
            16..=144 => {
```